### DISCUSSION TITLE
Debugging under MS Remote - SSH

### DISCUSSION THREAD

**Discussion by NerdFan** [vscode-abl/vscode-abl#27] [General]
Hi,

I'm not able to debug remote sessions.
Looks like ABL process is not started when debugging or wrong port opened.
I can run my code with "ABL: Run with _progres" - it means that OE part more or less configured.

Is there any ideas what can be wrong?

part of /tmp/vscode-debug-adapter.txt
```
[2022-09-03T00:17:27.786+0300] [INFO] eu.rssw.openedge.ls.a.h - Command line: [***/pro-11.7.5/bin/_progres, -basekey, INI, -ininame, /***/.vscode-server/extensions/riversidesoftware.openedge-abl-lsp-1.3.7/resources/abl-src/empty.ini, -p, /***/.vscode-server/extensions/riversidesoftware.openedge-abl-lsp-1.3.7/resources/abl-src/dynrun.p, -param, /***/repos/.builder/debug-cfg.json, -clientlog, /***/repos/.builder/debug-session.log, -debugReady, 9999]
[2022-09-03T00:17:27.878+0300] [ERROR] eu.rssw.openedge.dap.a - Unable to open sockets
java.net.ConnectException: Connection refused (Connection refused)
        at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
```

**Comment by gquerret** [comment]
Do you see anything interesting in `.builder/debug-session.log` ?

**Comment by NerdFan** [comment]
There is no debug-session.log created.
This is reason why I also assumed that debugging session did not start at all.

**Comment by NerdFan** [comment]
There is some "progress" on progress debugging.

1. I took execution parameters from log and started debug process manually:
```
/***/pro-11.7.5/bin/_progres -basekey INI -ininame /***/.vscode-server/extensions/riversidesoftware.openedge-abl-lsp-1.3.7/resources/abl-src/empty.ini -p /***/.vscode-server/extensions/riversidesoftware.openedge-abl-lsp-1.3.7/resources/abl-src/dynrun.p -param /***/vscode_test/.builder/debug-cfg.json -clientlog /***/vscode_test/.builder/debug-session.log -debugReady 3099
```

it gave an error - because of that probably it did not start debugging by self:
`The -s parameter requires a numeric argument. (11996)`

Also please note debug port - decision is not in sync to decide is it local or remote debugging -  here it expects remote (3099) while UI expects local and uses 9999

2. removed ini part and changed port:
```
/***/pro-11.7.5/bin/_progres -p /***/.vscode-server/extensions/riversidesoftware.openedge-abl-lsp-1.3.7/resources/abl-src/dynrun.p -param /***/vscode_test/.builder/debug-cfg.json -clientlog /***/vscode_test/.builder/debug-session.log -debugReady 9999
```

Got warning dialog on terminal - adding 11725 to swl.txt and -swl swl.txt helped to get rid of that:
`Use port 9999 for the Debugger to connect to. (11725)`

3. Run debug session from VSCode - it took manually started debug session, jumped over code and breakpoints till pause.
After pause it was able to debug rest of code.

**Comment by gquerret** [comment]
Interactive TTY sessions are not correctly started from the Debug Adapter (starting `_progres` from a background process does not allocate terminal, and the session stops with `Redirection or piping of stdin or stdout is allowed only with -b. (156)`). This will be investigated in the future. In the mean time, you can start your session with -debugReady, wait for message 11725, and then attach the debugger to it.

**Comment by Hanna19** [comment]
Hi,
I'm wondering if there is  a solution to the above problem. I tried to get Interactive TTY session working but all my attempts failed with error 156.
Hanna

**Comment by gquerret** [comment]
Start the remote application using [VS Code tasks](https://code.visualstudio.com/docs/editor/tasks) with `-debugReady` parameter, then attach the debugger to it.

**Comment by Hanna19** [comment]
Thank you, yes, I know. Using debugging with the request to launch is often more convenient way to debug one/few programs. Are there any plans to get it working?

**Reply by gquerret** [reply]
There are lots of items with higher priority.
